### PR TITLE
Include context in empty GraphQL errors

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- surface HTTP status, operation name, and response snippet when GraphQL responses omit both data and errors
- carry the query title into response processing for clearer diagnostics
- test empty response error messaging

## Testing
- `make fmt`
- `make lint`
- `make test`

------
https://chatgpt.com/codex/tasks/task_e_68b07932972c832292a38001c20b2f9b

## Summary by Sourcery

Include detailed context in errors when GraphQL responses omit both data and errors

Enhancements:
- Include HTTP status, operation name, and response snippet in error messages for empty GraphQL responses
- Return HTTP status code alongside response body from request execution
- Update retry logic to recognize the new “Empty GraphQL response” error prefix

Tests:
- Add test to verify that empty GraphQL responses report status, operation name, and body snippet

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Consider encapsulating the (status, body) tuple into a dedicated HttpResponse struct for clearer semantics and easier future extension rather than using a raw tuple.
- Instead of using a substring of the query to identify the operation, try extracting and passing the actual GraphQL operationName for more accurate diagnostics.
- It might be helpful to include the HTTP status and operation identifier in your transcript/logging path as well, so that retries and responses have full context when debugging.

File: src/api/tests.rs [hunk 1]
```diff
@@ -124,6 +124,26 @@ async fn run_query_retries_missing_data() {
     let _ = join.await;
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs:128]: Has this now been resolved?

File: src/api/mod.rs [hunk 2]
```diff
@@ -373,14 +378,15 @@
         let payload_str =
             serde_json::to_string(&payload).expect("serialising GraphQL request payload");
         let ctx = snippet(&payload_str, 1024).boxed();
+        let operation = snippet(query.as_ref(), 64);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs]: Has this now been resolved?

File: src/api/tests.rs [hunk 3]
```diff
@@ -24,6 +24,10 @@ struct TestClient {
 }
 
 fn start_server(responses: Vec<String>) -> TestClient {
+    start_server_with_status(responses, StatusCode::OK)
+}
+
+fn start_server_with_status(responses: Vec<String>, status: StatusCode) -> TestClient {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs:30]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: src/api/tests.rs [hunk 4]
```diff
@@ -116,14 +176,161 @@ async fn run_query_retries_missing_data() {
...
+    join.abort();
+    let _ = join.await;
+
+    let body = captured.lock().expect("lock").to_string();
+    let v: Value = serde_json::from_str(&body).expect("json body");
+    assert_eq!(
+        v.get("operationName").and_then(Value::as_str),
+        Some("RetryOp"),
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs:228]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai pause

File: src/api/tests.rs [hunk 5]
```diff
@@ -55,13 +53,109 @@ fn start_server(responses: Vec<String>) -> TestClient {
...
+            Ok::<_, Infallible>(
+                Response::builder()
+                    .status(status)
+                    .header("Content-Type", content_type)
+                    .body(Body::from(body))
+                    .expect("response"),
+            )
+        }
+    };
+    let (client, join, _) = create_test_server(handler);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs:91]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: src/api/mod.rs [hunk 6]
```diff
@@ -142,18 +148,40 @@ fn snippet(text: &str, max: usize) -> String {
...
+                .split(|c: char| c.is_whitespace() || c == '(' || c == '{')
+                .next()
+                .filter(|s| !s.is_empty());
+            if let Some(name) = name {
+                return Some(name);
+            }
+        }
+    }
+    None
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: src/api/tests.rs [hunk 7]
```diff
@@ -59,9 +57,71 @@
...
     TestClient { client, join }
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: src/api/tests.rs [hunk 8]
```diff
@@ -116,14 +176,162 @@
...
+    let (client, join, hits) = start_server_scripted(script);
+    let result: Value = client
+        .run_query("query HtmlRetry { __typename }", serde_json::json!({}))
+        .await
+        .expect("success after retry");
+    assert_eq!(result, serde_json::json!({"x": 1}));
+    assert!(hits.load(Ordering::SeqCst) >= 2, "expected at least 2 hits");
+    join.abort();
+    let _ = join.await;
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/tests.rs:261]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.

File: src/api/mod.rs [hunk 9]
```diff
@@ -261,13 +292,17 @@ impl GraphQLClient {
...
-            VkError::BadResponse(msg) => msg.starts_with("Missing data in response"),
+            VkError::RequestContext { .. }
+            | VkError::Request(_)
+            | VkError::EmptyResponse { .. } => true,
+            VkError::BadResponseSerde {
+                status, snippet, ..
+            } => *status >= 500 || *status == 429 || snippet.trim_start().starts_with('<'),
             _ => false,
         }
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions in a tilde code-fenced block.
